File: 백준/Gold/2293. 동전 1/동전 1.py
import sys
input = sys.stdin.readline

num, coin = map(int, input().split())
cost = []
for _ in range(num):
    cost.append(int(input()))
    
dp = [0 for _ in range(coin + 1)]
dp[0] = 1
for cn in cost:
    for k in range(cn, coin + 1):
        dp[k] += dp[k-cn]

print(dp[coin])

File: 백준/Gold/2293. 동전 1/동전 1.cc
#include <iostream>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    
    int num, coin;
    cin >> num >> coin;
    int cost[100];
    for(int i = 0; i < num; i++)
        cin >> cost[i];
    
    int dp[10001] = {0};
    dp[0] = 1;
    for(int i = 0; i < num; i++){
        for(int j = cost[i]; j < coin+1; j++) {
            dp[j] += dp[j-cost[i]];
        }
    }

    cout << dp[coin] << endl;
    
    return 0;
}